File: Linear/Dynamic/Linked List/Circular Linked List/01_Inserting/End/Inserting.js
class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

function insertAtEnd(tail, x) {
    const newNode = new Node(x);

    if (tail === null) {
        newNode.next = newNode;
        return newNode;
    }

    newNode.next = tail.next;
    tail.next = newNode;

    tail = newNode;

    return tail;
}

function printList(last) {
    if (last === null)
        return;

    let head = last.next;
    let temp = head;

    do {
        process.stdout.write(temp.data.toString())
        temp = temp.next;
        if (temp !== head)
            process.stdout.write(" -> ");
    } while (temp !== head);

    console.log();
}

// Create circular linked list: 2 -> 3 -> 4
let first = new Node(2);
first.next = new Node(3);
first.next.next = new Node(4);

let last = first.next.next;
last.next = first;

console.log("Linked List before insertion");
printList(last);

const x = 5
last = insertAtEnd(last, x);

console.log("Linked List after insertion");
printList(last);

File: Linear/Dynamic/Linked List/Circular Linked List/01_Inserting/End/Inserting.java
class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class InsertingEnd {
    public static Node insertAtEnd(Node tail, int x) {
        Node newNode = new Node(x);

        if (tail == null) {
            newNode.next = newNode;
            return newNode;
        }

        newNode.next = tail.next;
        tail.next = newNode;

        tail = newNode;

        return tail;
    }

    public static void printList(Node last) {
        if (last == null)
        return;

        Node head = last.next;
        Node temp = head;

        do {
            System.out.print(temp.data);
            temp = temp.next;
            if (temp != head)
            System.out.print(" -> ");
        } while (temp != head);

        System.out.println();
    }

    public static void main(String[] args) {
        // Create circular linked list: 2 -> 3 -> 4
        Node first = new Node(2);
        first.next = new Node(3);
        first.next.next = new Node(4);

        Node last = first.next.next;
        last.next = first;

        System.out.println("Linked List before insertion");
        printList(last);

        int x = 5;
        last = insertAtEnd(last, x);

        System.out.println("Linked List after insertion");
        printList(last);
    }
}